### PR TITLE
Structure Pool Callback

### PR DESCRIPTION
**Pull request by frqnny**
Basically redos #1389 Even though it's exactly the same except FabricStructurePool is a record.
This PR adds a callback to allow players to modify structure pools. 

I was going to make it different but I decided not too- I was thinking of using DynamicRegistryLoadingEvent or whatever it's called, but that loads before so nah. So this is basically just retargetted to 1.17. 

Anyway, the important part is that this is data-pack safe. Repurposed Structures, for example, uses structure pool in json instead of using builtin and they show up using the testmod and RS together. I guess you could test this by making a RS addon? The best use case, though, will be to allow people to add houses to villages. Say you add a profession, you will need a house for that profession, this is the way you add it.

### CODE REVIEW

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
Checkstyle is too old. It needs to be bumped to 8.40+ to support records (latest as of comment is 8.44)

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (liach) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]:
```suggestion

		if (pool.getElementCounts() instanceof ArrayList) {
```

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (liach) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]:
```suggestion
```

**Review by liach — COMMENTED** [review]
checkstyle

**Comment by frqnny** [comment]
ill fix soon don’t worry

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (maityyy) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]:
```java
(pool).getElementCounts()
```
???

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java
Comment (maityyy) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java]: can be simplified btw
```java
if (pool instanceof StructurePool p) {
    StructurePoolAddCallback.EVENT.invoker().onAdd(new FabricStructurePool(p));
}
```

Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java]: thanks for these catches

**Comment by frqnny** [comment]
Ooh, got an idea. gonna expose the identifier so they don't have to do #underlying

**Comment by frqnny** [comment]
Updated javadocs too. Should be great now :D

**Comment by Technici4n** [comment]
Looks really nice and concise. Would there be a way to test this, such as a testmod ?

**Comment by frqnny** [comment]
This PR is based off Structurized, a mod by Draylar and mantained by me
Here is slightly outdated TacoCraft adding corn farms to plains villages example 
https://github.com/frqnny/TacoCraft/blob/0491e061b467949ef1615dbef48af3cbed3ade02/src/main/java/io/github/frqnny/tacocraft/init/ModGen.java#L40
Here is golems galore adding its golems to Iron Golem pool to make its golems spawn in villages without breaking village iron farms. 
https://github.com/frqnny/golemsgalore/blob/c0fa0f5f5f017f7ac226ef92b5dc8749c6789fca/src/main/java/io/github/frqnny/golemsgalore/GolemsGalore.java#L39
the test mod and the javadocs have some examples

**Review by Technici4n — CHANGES_REQUESTED** [review]
Please add a testmod.

Every time I consider reviewing this PR I am reminded that I'll have to `gradlew publishToMavenLocal` and clone another mod and test it there. This is too much effort, especially if an update breaks this addition and we need to quickly test it.

The testmod should ideally add a structure, validate that the callback gets invoked, and possibly even inspect the structure pool after server start to make sure the addition succeeded.

**Comment by frqnny** [comment]
Will do sometime this week. 


EDIT:

it is now 9/28, sorry I haven't done it. I can't promise anything so this PR is on hold until I'm able to work on it again.

**Comment by frqnny** [comment]
Just find a desert village and this should appear when testing. Might take a while, using /locatebiome, /locate and /tp (10k blocks away to find new villages)

**Comment by frqnny** [comment]
image

File: build.gradle
Comment (modmuss50) [inline build.gradle]: Is there a reason for this update?

File: fabric-structure-api-v1/build.gradle
Comment (modmuss50) [inline fabric-structure-api-v1/build.gradle]: New line please.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (modmuss50) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]: Is a record really the right choice here? An interface would be better imo:

- A record (and interface) shouldnt really contain logic
- It makes it easier for someone else (or another module) to re-impliment the API.
- It makes the API surface clear, I dont think its too bad here but would be better as a simple interface.

The impl could still be a record (if it makes sense) but the exposed API should really be an interface imo.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (modmuss50) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]: This logic should be "hidden" away in an impl somewhere, moving to an interface will solve this.

**Review by modmuss50 — CHANGES_REQUESTED** [review]

File: build.gradle
Comment (frqnny) [reply build.gradle]: it was a merge conflict that I resolved. it'd be complaining about it if I didn't, IDK why the former red is showing that it is 8.43 since I was just updating it to match 9.1

https://github.com/FabricMC/fabric/blob/fa34ed2c10c399ed6c973a56f3fda31cd47811f6/build.gradle#L142

**Comment by frqnny** [comment]
I've gone ahead and updated Structurized to 1.18 and to match with this PR's latest changes. It required to real changes from 1.17 to 1.18, so the big thing is that it's an exact replica of this PR. There is still our new testing material, but I will also post tommorow-ish how we use this mod with my own mods in 1.18 to prove that this PR works. (Structurized 1.3.0 worked in 1.17, Structurized 1.4 works in 1.18) https://github.com/frqnny/structurized

**Comment by maityyy** [comment]
Is this event needed for anything else? I think it's better to convert all the events needed to modify something into something like the BiomeModification API, which is much more convenient (probably).

**Comment by frqnny** [comment]
The event can be used for manipulating villages, bastions, and other structure pools. The problem is that IMO something like BiomeModification would  be over-engineering the entire thing. We only do one event, which should be simple enough for modders (no more than 7 lines). Something similar to Biome Modification is a bit too much for something like a StructurePool as we can just give them the FabricStructurePool in an event and get done with it.

**Comment by frqnny** [comment]
**All I want for Christmas... is**

~~YOUUUUUU~~ this pr to get reviews


jokes aside just a small subtle reminder to take your time and review when you are open, thanks devs for the amazing work and Merry Christmas.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]:
```suggestion
	 * @param weight  Minecraft handles weight by adding it that amount of times into a list
```

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java:29]: What details? It has the exact same description excluding this line

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java:26]: Maybe `@ApiStatus.NonExtendable`?

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/StructurePoolAddCallback.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/StructurePoolAddCallback.java]: This example is a bit ugly, I'd just add a field with that identifier and compare that instead of the string

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/StructurePoolAddCallback.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/StructurePoolAddCallback.java:26]: Out of scope for this PR, but a possible future improvement would be a method in `FabricStructurePool` to check whether it's from a data pack (some mods might not want to modify user data)

File: fabric-structure-api-v1/src/main/resources/fabric-structure-api-v1.accesswidener
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/resources/fabric-structure-api-v1.accesswidener]: No need for this if you mark your `@Accessor` setter with `@Mutable` in `StructurePoolAccessor`

File: fabric-structure-api-v1/src/testmod/java/net/fabricmc/fabric/test/structure/StructureTest.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/testmod/java/net/fabricmc/fabric/test/structure/StructureTest.java]:
```suggestion
		StructurePoolAddCallback.EVENT.register(structurePool -> LOGGER.info("Structure pool {} added", structurePool.getId()));
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (haykam821) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java]: I'm not sure this is a good parameter description since this is an implementation detail.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java
Comment (deirn) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java]: Maybe add a check that `weight > 0`?

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java
Comment (deirn) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java:31]: Is the explicit target necessary? :thinking:

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java:29]: Ok, I'll rephrase it a bit.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java:26]: Doesn't that mean it shouldn't be implemented?
modmuss requested that the class and impl be split to also include an interface in case someone wants to use it.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java:31]: It doesn't hurt to have them. Unless there's some strict formatting rule about this, IMO it is better to have them as they make it extra clear what field they are supposed to access.

**Comment by frqnny** [comment]
Thanks for the reviews, this PR has been updated accordingly. Hope to get this merged soon!

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java
Comment (Juuxel) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java]: Should be on the setter I think, but mixin might apply the `final` removal here too (even though it's a bit weird)

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (Juuxel) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java:26]: This API interface + impl class pattern is usually used so that we can change the impl's structure without breaking anything (or even move it to another module). I don't think users gain anything in particular from implementing this itf, but I guess it can stay how it is without the anno.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java]: Okay, I'll try to see what happens when I remove it.

Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructurePoolAccessor.java]: Seems to work fine, now it is just on the setter.

**Review by Player3324 — APPROVED** [review]
LGTM with the pending review comments resolved. Thanks!

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java
Comment (deirn) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java]: Since the `DynamicRegistryManager.Info<E> info` isn't used in this inject, you can replace it with `@Coerce Object info` instead, thus making the AW not needed.

[Coerce javadoc](https://jenkins.liteloader.com/job/Mixin/javadoc/org/spongepowered/asm/mixin/injection/Coerce.html)

**Review by Juuxel — APPROVED** [review]

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/DynamicRegistryManagerMixin.java]: That's cool, thanks for the heads up! Glad we could remove the AW.

**Review by deirn — APPROVED** [review]

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java
Comment (modmuss50) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructurePool.java:26]: This is fine, I dont believe we use that on other similar interfaces.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java
Comment (modmuss50) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java]: I think it would be best to make the list immutable before setting it back again. Would remove the need to check if its an arrary list.

If peformace is a concern (I dont think it is here?) setting the array back to an immutable one after the callback has been invoked would also be good.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java
Comment (modmuss50) [inline fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java]: I would rename this to `poolAccessor` or something, as its hiding the `pool` field.

**Review by modmuss50 — APPROVED** [review]
API looks good, just a few minor things I'd change in the impl.

I will add this to last call. 👍

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java]: All right, now it just temporarily replaces whatever collection with an array list, then sets it back as an Immutable List.

File: fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java
Comment (frqnny) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java]: done.

Comment (modmuss50) [reply fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructurePoolImpl.java]: Looks great, thanks 👍

**Comment by frqnny** [comment]
As of the latest commit, the callback is working, sadly a cave ate our cactus farm :(

**Pull request merged**
